core/primitives/src/lib.rs

```diff
@@ -286,3 +286,25 @@ impl From<LogLevel> for log::Level {
 		}
 	}
 }
+
+/// Converts the given value into an `u64` by following the Substrate calling convention.
+///
+/// When Substrate calls into Wasm it expects a fixed signature for functions exported
+/// from the Wasm blob. The return value of this signature is always a `u64` in little endian.
+/// This `u64` stores the pointer to the encoded return value and the length of this encoded value.
+/// The first `32bit` are reserved for the pointer and the last `32bit` for the length.
+#[cfg(not(feature = "std"))]
+pub fn to_substrate_wasm_fn_return_value(value: &impl Encode) -> u64 {
+	let encoded = value.encode();
+
+	let ptr = encoded.as_ptr() as u64;
+	let length = encoded.len() as u64;
+	let res = ptr | (length << 32);
+
+	// Leak the output vector to avoid it being freed.
+	// This is fine in a WASM context since the heap
+	// will be discarded after the call.
+	rstd::mem::forget(encoded);
+
+	res.to_le()
```

Comment:
As mentioned above this doesn't make a lot of sense:

1. This is a function that is supposed to be used only in wasm thus conversion to little endian would do nothing
2. If this code were used on a big-endian machine, this would be incorrect since the bytes would be swapped.

Reply:
Wasm is always little endian. So yes, the conversion is somewhat useless, but will be optimized away anyway. 

Now on the host machine the current code works, because we are running it on little endian machines as well, but would you run it on a big endian machine, it would fail.

This representation is just for communicating between the wasm runtime and the host. Think of networking, in networking you store data in big endian. You need a fixed endianess, to know how to decode the data on the other side.